server/bleep/src/remotes.rs

```diff
@@ -199,33 +200,88 @@ pub(crate) fn gather_repo_roots(
         })
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug)]
-pub(crate) enum BackendCredential {
-    Github(github::Auth),
+struct BackendEntry {
+    inner: BackendCredential,
+    updated: flume::Receiver<()>,
+    updated_tx: flume::Sender<()>,
 }
 
-impl BackendCredential {
-    pub(crate) async fn validate(&self) -> Result<()> {
-        let BackendCredential::Github(auth) = self;
+impl From<BackendCredential> for BackendEntry {
+    fn from(inner: BackendCredential) -> Self {
+        let (updated_tx, updated) = flume::unbounded();
+        BackendEntry {
+            inner,
+            updated_tx,
+            updated,
+        }
+    }
+}
 
-        let client = auth.client()?;
+#[derive(Clone)]
+pub struct Backends {
+    backends: Arc<DashMap<Backend, BackendEntry>>,
+}
 
-        match client.current().user().await {
-            Ok(_) => {}
-            Err(e @ octocrab::Error::GitHub { .. }) => {
-                warn!(?e, "failed to validate GitHub token");
-                return Err(e)?;
-            }
-            Err(e) => {
-                // Don't return an error here - we want to swallow failure and try again on the
-                // next poll.
-                error!(?e, "failed to make GitHub user request");
-            }
+impl From<DashMap<Backend, BackendCredential>> for Backends {
+    fn from(value: DashMap<Backend, BackendCredential>) -> Self {
+        Self {
+            backends: Arc::new(value.into_iter().map(|(k, v)| (k, v.into())).collect()),
         }
+    }
+}
+
+impl Backends {
+    pub(crate) fn for_repo(&self, repo: &RepoRef) -> Option<BackendCredential> {
+        let Some(handle) = self.backends.get(&repo.backend()) else {
+		return None;
+	    };
+
+        Some(handle.value().inner.clone())
+    }
+
+    pub(crate) fn remove(&self, backend: impl Borrow<Backend>) -> Option<BackendCredential> {
+        self.backends.remove(backend.borrow()).map(|(_, v)| v.inner)
+    }
+
+    pub(crate) fn github(&self) -> Option<github::State> {
+        let Some(handle) = self.backends.get(&Backend::Github) else {
+	    return None;
+	};
 
-        Ok(())
+        let BackendCredential::Github(ref github) = handle.value().inner;
+        Some(github.clone())
     }
 
+    pub(crate) fn set_github(&self, gh: github::State) {
```

Comment:
Could the `set_` and `update_` methods be made generic?

Reply:
Realistically, I don't think they can be, or at least not without larger changes. My aim here was to clean up the call sites. I don't really know how this will scale when we start adding other backends, but I'd probably leave that problem for our future selves to deal with.